File: src/main/java/org/cubexmc/railway/gui/ItemBuilder.java
package org.cubexmc.railway.gui;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

/**
 * Utility class for simplifying ItemStack creation
 */
public class ItemBuilder {
    private final ItemStack item;
    private final ItemMeta meta;

    public ItemBuilder(Material material) {
        this.item = new ItemStack(material);
        this.meta = item.getItemMeta();
    }

    public ItemBuilder(Material material, int amount) {
        this.item = new ItemStack(material, amount);
        this.meta = item.getItemMeta();
    }

    public ItemBuilder name(String name) {
        if (meta != null) {
            meta.setDisplayName(ChatColor.translateAlternateColorCodes('&', name));
        }
        return this;
    }

    public ItemBuilder lore(String... lore) {
        if (meta != null) {
            List<String> loreList = Arrays.stream(lore)
                    .map(s -> ChatColor.translateAlternateColorCodes('&', s))
                    .collect(Collectors.toList());
            meta.setLore(loreList);
        }
        return this;
    }

    public ItemBuilder lore(List<String> lore) {
        if (meta != null) {
            List<String> loreList = lore.stream()
                    .map(s -> ChatColor.translateAlternateColorCodes('&', s))
                    .collect(Collectors.toList());
            meta.setLore(loreList);
        }
        return this;
    }

    public ItemBuilder glow() {
        if (meta != null) {
            meta.addEnchant(Enchantment.UNBREAKING, 1, true);
            meta.addItemFlags(ItemFlag.HIDE_ENCHANTS);
        }
        return this;
    }

    public ItemBuilder hideAttributes() {
        if (meta != null) {
            meta.addItemFlags(ItemFlag.HIDE_ATTRIBUTES);
        }
        return this;
    }

    public ItemStack build() {
        if (meta != null) {
            item.setItemMeta(meta);
        }
        return item;
    }
}

File: src/main/java/org/cubexmc/metro/gui/ItemBuilder.kt
package org.cubexmc.metro.gui

import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import org.cubexmc.gui.ItemBuilder as SharedItemBuilder
import org.cubexmc.gui.TextStyler
import org.cubexmc.metro.util.ColorUtil

/**
 * 简化物品创建的工具类。
 *
 * 构造逻辑本身在共享模块 `cubex-gui` 的 [SharedItemBuilder]；这里只把它绑定到本插件的颜色管线
 * （legacy `&` + `&#RRGGBB`），让 GUI 的 90 多个调用点保持原样。
 */
class ItemBuilder @JvmOverloads constructor(material: Material, amount: Int = 1) {
    private val delegate = SharedItemBuilder(material, amount, STYLER)

    fun name(name: String): ItemBuilder = apply { delegate.name(name) }

    fun lore(vararg lore: String): ItemBuilder = apply { delegate.lore(*lore) }

    fun lore(lore: List<String>): ItemBuilder = apply { delegate.lore(lore) }

    fun glow(): ItemBuilder = apply { delegate.glow() }

    fun hideAttributes(): ItemBuilder = apply { delegate.hideAttributes() }

    fun build(): ItemStack {
        return delegate.build()
    }

    private companion object {
        // colorize 的返回值可空；退回原串比抛 NPE 更符合"按钮至少画得出来"的预期。
        private val STYLER = TextStyler { input -> ColorUtil.colorize(input) ?: input }
    }
}
